src/Storages/WindowView/StorageWindowView.cpp

```diff
@@ -0,0 +1,1382 @@
+
+    struct ReplaceFunctionWindowMatcher
+    {
+        using Visitor = InDepthNodeVisitor<ReplaceFunctionWindowMatcher, true>;
+
+        struct Data{};
+
+        static bool needChildVisit(ASTPtr &, const ASTPtr &) { return true; }
+
+        static void visit(ASTPtr & ast, Data &)
+        {
+            if (auto * t = ast->as<ASTFunction>())
+            {
+                if (t->name == "HOP" || t->name == "TUMBLE")
+                    t->name = "WINDOW_ID";
+            }
+        }
+    };
+
+    class ToIdentifierMatcher
+    {
+    public:
+        using Visitor = InDepthNodeVisitor<ToIdentifierMatcher, true>;
+
+        struct Data
+        {
+            String window_id_name;
+            String window_id_alias;
+            Aliases * aliases;
+        };
+
+        static bool needChildVisit(ASTPtr &, const ASTPtr &) { return true; }
+
+        static void visit(ASTPtr & ast, Data & data)
+        {
+            if (const auto * t = ast->as<ASTFunction>())
+                visit(*t, ast, data);
+            if (const auto * t = ast->as<ASTIdentifier>())
+                visit(*t, ast, data);
+        }
+
+    private:
+        static void visit(const ASTFunction & node, ASTPtr & node_ptr, Data &)
+        {
+            if (node.name == "tuple")
+                return;
+            else
+                node_ptr = std::make_shared<ASTIdentifier>(node.getColumnName());
+        }
+
+        static void visit(const ASTIdentifier & node, ASTPtr & node_ptr, Data & data)
+        {
+            if (node.getColumnName() == data.window_id_alias)
+            {
+                if (auto identifier = std::dynamic_pointer_cast<ASTIdentifier>(node_ptr))
+                    identifier->setShortName(data.window_id_name);
+            }
+        }
+    };
+
+    IntervalKind strToIntervalKind(const String& interval_str)
+    {
+        if (interval_str == "Second")
+            return IntervalKind::Second;
+        else if (interval_str == "Minute")
+            return IntervalKind::Minute;
+        else if (interval_str == "Hour")
+            return IntervalKind::Hour;
+        else if (interval_str == "Day")
+            return IntervalKind::Day;
+        else if (interval_str == "Week")
+            return IntervalKind::Week;
+        else if (interval_str == "Month")
+            return IntervalKind::Month;
+        else if (interval_str == "Quarter")
+            return IntervalKind::Quarter;
+        else if (interval_str == "Year")
+            return IntervalKind::Year;
+        __builtin_unreachable();
+    }
+
+    void extractWindowArgument(const ASTPtr & ast, IntervalKind::Kind & kind, Int64 & num_units, String err_msg)
+    {
+        const auto * arg = ast->as<ASTFunction>();
+        if (!arg || !startsWith(arg->name, "toInterval"))
+            throw Exception(err_msg, ErrorCodes::ILLEGAL_TYPE_OF_ARGUMENT);
+
+        kind = strToIntervalKind(arg->name.substr(10)); /// TODO: why substr?
```

Comment:
Why do we need `substr` here?
To me it seems it is not needed.

Reply:
`substr` is used to discard the "toInterval" in the string "toIntervalXXX". For example, convert `toIntervalSecond` to `Second`

Reply:
ah, of course.